pennylane/interfaces/jax_jit.py

```diff
@@ -91,174 +105,114 @@ def _jac_shape_dtype_struct(tape: "qml.tape.QuantumScript", device: "qml.Device"
-    return jnp.zeros(shape_dtype.shape, shape_dtype.dtype)
-
+def _pytree_shape_dtype_struct(pytree):
+    """Creates a shape structure that matches the types and shapes for the provided pytree."""
+    leaves, struct = jax.tree_util.tree_flatten(pytree)
+    new_leaves = [jax.ShapeDtypeStruct(jnp.shape(l), l.dtype) for l in leaves]
+    return jax.tree_util.tree_unflatten(struct, new_leaves)
 
-def _switched_jacobian(tape, device, original_trainable_parameters, jacs):
-    """Adds in jacobians with of all zeros for parameters that had zero tangent.
-
-    Used in ``execute_wrapper_jvp`` for grad_on_execution=True with device derivatives.
-
-    Note that this adds an additional nesting dimension to ``jacs``, with one jacobian
-    for each original trainable parameter. I am unsure why this works.
-
-    >>> dev = qml.devices.DefaultQubit()
-    >>> config = qml.devices.ExecutionConfig(gradient_method="adjoint")
-    >>> tape = qml.tape.QuantumTape([qml.RY(1.0, 0), qml.RX(0.6, 0), qml.RX(0.7, 0)], [qml.expval(qml.PauliZ(0))])
-    >>> tape.trainable_params = [0, 2]
-    >>> jac = dev.compute_derivatives(tape, config)
-    >>> jac
-    (array(-0.2250925), array(-0.52061271))
-    >>> _switched_jacobian(tape, dev, [0,1,2], jac)
-    ((array(-0.2250925), array(-0.52061271)),
-    (Array(0., dtype=float64), Array(0., dtype=float64)),
-    (array(-0.2250925), array(-0.52061271)))
 
+def _execute_wrapper(params, tapes, execute_fn, _, device) -> ResultBatch:
     """
-    intermediate_jacs = []
-
-    shape_dtype = _jac_shape_dtype_struct(tape, device)
-
-    for param_idx in original_trainable_parameters:
-        if param_idx in tape.trainable_params:
-            p_jac = jacs
-        else:
-            p_jac = _empty_jacs_for_shape(shape_dtype)
-        intermediate_jacs.append(p_jac)
-    return tuple(intermediate_jacs)
-
+    Execute tapes using a pure-callback.
+    """
+    shape_dtype_structs = tuple(_result_shape_dtype_struct(t, device) for t in tapes.vals)
+
+    def pure_callback_wrapper(p):
+        new_tapes = _set_parameters_on_copy(tapes.vals, p)
+        res = tuple(execute_fn(new_tapes))
+        # When executed under `jax.vmap` the `result_shapes_dtypes` will contain
+        # the shape without the vmap dimensions, while the function here will be
+        # executed with objects containing the vmap dimensions. So res[i].ndim
+        # will have an extra dimension for every `jax.vmap` wrapping this execution.
+        #
+        # The execute_fn will return an object with shape `(n_observables, batches)`
+        # but the default behaviour for `jax.pure_callback` is to add the extra
+        # dimension at the beginning, so `(batches, n_observables)`. So in here
+        # we detect with the heuristic above if we are executing under vmap and we
+        # swap the order in that case.
+        return jax.tree_map(lambda r, s: r.T if r.ndim > s.ndim else r, res, shape_dtype_structs)

```

Comment:
Good catch :sweat_smile: 
Just to make sure: Does this transpose do the right reordering for multiple `vmap`s as well? :)

Reply:
Copy and paste from the previous version.  I don't technically understand the change.